### DISCUSSION TITLE
关于 `cus` 编程接口的一些问题。

### DISCUSSION THREAD

**Discussion by xkwxdyy** [Sophanatprime/cus#50] [General]
主要想咨询一些关于编程接口的细节和疑问。

**Comment by xkwxdyy** [comment]
## 关于展开两次

我早上看到这个 `cus_exp_args:Nd`
![image](https://github.com/Sophanatprime/cus/assets/87134063/4cb9b48a-c532-4afe-b4a1-a12d3c0269b8)


我很好奇为什么展开两次会变成这个效果。我本来想用 `unravel` 包来看效果的，结果发现
```latex
\documentclass{ctexart}
\makeatletter
\@ifundefined{input@path}{\let\input@path\@empty}{}
\g@addto@macro\input@path{{./module}{./library}{./lt3ekeys}{./cus-doc}{./cus-doc-aux}}
\makeatother
\usepackage{cus}
\usepackage{unravel}


\begin{document}

\ExplSyntaxOn
\unravel{
  \prg_replicate:nn {5}{\CusTeX}
}
\ExplSyntaxOff

\end{document}
```
展开过程有几十步，前几步是：
```latex
|> \prg_replicate:nn {5}{\CusTeX }

[===== Step 1 =====] \prg_replicate:nn = \long macro:#1->\exp:w \exp...
|| 
|> \exp:w \exp_after:wN \__prg_replicate_first:N \int_value:w \int_eval:n
|> {5}\cs_end: {\CusTeX }

[===== Step 2 =====] \exp:w = \romannumeral
|| \exp:w 
|> \exp_after:wN \__prg_replicate_first:N \int_value:w \int_eval:n {5}\cs_end:
|> {\CusTeX }

[===== Step 3 =====] \exp_after:wN \__prg_replicate_first:N 
|| \exp:w 
|| \exp_after:wN \__prg_replicate_first:N 
|> \int_value:w \int_eval:n {5}\cs_end: {\CusTeX }

[===== Step 4 =====] \int_value:w = \number
|| \exp:w 
|| \exp_after:wN \__prg_replicate_first:N 
|| \int_value:w 
|> \int_eval:n {5}\cs_end: {\CusTeX }

[===== Step 5 =====] \int_eval:n = \long macro:#1->\tex_the:D \__int...
|| \exp:w 
|| \exp_after:wN \__prg_replicate_first:N 
|| \int_value:w 
|> \tex_the:D \__int_eval:w 5\__int_eval_end: \cs_end: {\CusTeX }

[===== Step 6 =====] \tex_the:D = \the
|| \exp:w 
|| \exp_after:wN \__prg_replicate_first:N 
|| \int_value:w 
|| \tex_the:D 
|> \__int_eval:w 5\__int_eval_end: \cs_end: {\CusTeX }
```
让我更迷惑了，这个“展开两次”的过程是怎么样的呢？能麻烦您解释一下吗

**Reply by Sophanatprime** [reply]
`\exp_args:No` 这种“展开一次”是执行一次 `\exp_after:wN <f> <tokens_a>`，假设 `<tokens_a>` 的展开结果为 `<tokens_b>`，然后再对 `<tokens_b>` 执行一次相同的操作，结果为 `<tokens_c>`，那么 `<tokens_c>` 就是 `<tokens_a>` “展开两次”的结果。与 `<tokens_a>` `<tokens_b>` 内部是怎么展开的、展开了多少次无关。

`unravel` 就是告诉你内部每一步是怎么做的。
现在 `\prg_replicate:nn ...` “展开一次”变成了 `\exp:w ...`，然后再展开 `\exp:w ...` 一次就得到结果，从外界看，只进行了两次展开操作。

很多“展开两次得到结果”（exactly two expansions are needed / require two expansions）的命令就是使用了 `\exp:w ... \exp_end:` 这样的构造，`\exp:w` 和 `\exp_end:` 中间的 tokens 完全展开为空，结果放在 `\exp_end:` 的后面，当展开完 `\exp_end:` 之后，`\exp:w` 结束，`\exp_end:` 消失，只留下了 `\exp_end:` 后面的结果。

**Reply by xkwxdyy** [reply]
谢谢，有一些了解了，但太深的也不太懂。我刚试了一下这个例子：
```latex
\exp_args:No \tl_to_str:n 
  { \CusTeX }

\par
\cus_exp_args:Nd \tl_to_str:n 
  { \CusTeX }

\par
\exp_args:Nx \tl_to_str:n 
  { \CusTeX }
```
![image](https://github.com/Sophanatprime/cus/assets/87134063/09fb634a-4ea2-4906-b89a-58c3b5c4c774)

为什么展开 1 次和 2 次的结果是一样的呢，`\TeX` 为什么没被展开呢？但是可以被 `x` 展开

**Reply by Sophanatprime** [reply]
`o` 是展开一次，它展开第一个记号。第二个例子在进行第二次展开时，遇到的记号为 `c`，不能展开，结果保持不变。
`x` `e` 是完全展开，它展开参数中所有能被展开的记号，`\TeX` 不是 `\protected`，所以会展开 `\TeX`，`\kern` 等是不能被展开的。

**Reply by xkwxdyy** [reply]
那可以理解为 `d` 其实就是两次 `o` 吗?

**Reply by Sophanatprime** [reply]
对

**Reply by xkwxdyy** [reply]
好的，感谢。
